src/WebJobs.Extensions.DurableTask.Analyzers/Analyzers/ActivityFunction/ArgumentAnalyzer.cs

```diff
@@ -13,28 +14,106 @@ public class ArgumentAnalyzer
+        public static readonly DiagnosticDescriptor InputNotUsedRule = new DiagnosticDescriptor(DiagnosticId, Title, InputNotUsedMessageFormat, Category, Severity, isEnabledByDefault: true, description: Description);
 
-        public static void ReportProblems(CompilationAnalysisContext context, IEnumerable<ActivityFunctionDefinition> availableFunctions, IEnumerable<ActivityFunctionCall> calledFunctions)
+        public static void ReportProblems(CompilationAnalysisContext context, SemanticModel semanticModel, IEnumerable<ActivityFunctionDefinition> availableFunctions, IEnumerable<ActivityFunctionCall> calledFunctions)
         {
-            foreach (var node in calledFunctions)
+            foreach (var activityInvocation in calledFunctions)
             {
-                if (availableFunctions.Where(x => x.FunctionName == node.Name).Any())
+                var functionDefinition = availableFunctions.Where(x => x.FunctionName == activityInvocation.Name).SingleOrDefault();
+                if (functionDefinition != null)
                 {
-                    var functionDefinition = availableFunctions.Where(x => x.FunctionName == node.Name).SingleOrDefault();
-                    if (functionDefinition.InputType != node.ParameterType)
+                    TryGetInvocationInputType(semanticModel, activityInvocation, out ITypeSymbol invocationInputType);
+                    TryGetDefinitionInputType(semanticModel, functionDefinition, out ITypeSymbol definitionInputType);
+
+                    var invocationTypeName = SyntaxNodeUtils.GetQualifiedTypeName(invocationInputType);
+                    var definitionTypeName = SyntaxNodeUtils.GetQualifiedTypeName(definitionInputType);
+
+                    if (definitionInputType == null)
                     {
-                        var diagnostic = Diagnostic.Create(Rule, node.ParameterNode.GetLocation(), node.Name, functionDefinition.InputType, node.ParameterType);
+                        var diagnostic = Diagnostic.Create(InputNotUsedRule, activityInvocation.ParameterNode.GetLocation(), activityInvocation.Name);
+
+                        context.ReportDiagnostic(diagnostic);
+                    }
+                    else if (!SyntaxNodeUtils.InputMatchesOrCompatibleType(invocationInputType, invocationTypeName, definitionInputType, definitionTypeName))
+                    {
+                        var diagnostic = Diagnostic.Create(Rule, activityInvocation.ParameterNode.GetLocation(), activityInvocation.Name, definitionTypeName, invocationTypeName);
 
                         context.ReportDiagnostic(diagnostic);
                     }
                 }
             }
         }
+
+        private static void TryGetInvocationInputType(SemanticModel semanticModel, ActivityFunctionCall activityInvocation, out ITypeSymbol invocationInputType)
```

Comment:
Should I change this to return ITypeSymbol or make this match the TryGet pattern with bool? I don't care, in this case, if it returns a null value. But I thought I should still make the method a TryGet just to let the caller know that it could return a null value, if I end up caring that it returns null in the future.